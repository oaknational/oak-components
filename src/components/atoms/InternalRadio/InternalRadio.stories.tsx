import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import {
  InternalRadio,
  InternalRadioHover,
  InternalRadioFocus,
  InternalRadioHoverFocus,
} from "./InternalRadio";

import { OakBox } from "@/components/atoms/OakBox";
import { colorArgTypes } from "@/storybook-helpers/colorStyleHelpers";
import { spacingArgTypes } from "@/storybook-helpers/spacingStyleHelpers";
import { borderArgTypes } from "@/storybook-helpers/borderStyleHelpers";
import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

console.log({ sizeArgTypes });

const meta: Meta<typeof InternalRadio> = {
  component: InternalRadio,
  tags: ["autodocs"],
  title: "components/atoms/InternalRadio",
  argTypes: {
    ...sizeArgTypes,
    ...borderArgTypes,
    ...colorArgTypes,
    ...spacingArgTypes,
    disabled: {
      control: "boolean",
    },
    $checkedBackground: colorArgTypes.$color,
  },
  parameters: {
    controls: {
      include: [
        "disabled",
        "defaultChecked",
        "$width",
        "$height",
        "$ba",
        "$borderRadius",
        "$borderColor",
        "$hoverBorderRadius",
        "$checkedBackground",
      ],
    },
  },
};
export default meta;

export const Default: StoryObj<typeof InternalRadio> = {
  render: (args) => <InternalRadio {...args} />,
  args: {
    id: "checkbox-test-default-1",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const Hover: StoryObj<typeof InternalRadioHover> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalRadioHover {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-2",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const Focus: StoryObj<typeof InternalRadioFocus> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalRadioFocus {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-3",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const HoverFocus: StoryObj<typeof InternalRadioHoverFocus> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalRadioHoverFocus {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-4",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};
