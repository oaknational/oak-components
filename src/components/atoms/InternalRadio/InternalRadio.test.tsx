import React, { createRef } from "react";
import "@testing-library/jest-dom";
import { fireEvent } from "@testing-library/react";

import { InternalRadio } from "./InternalRadio";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe("InternalRadio", () => {
  it("renders a radio", () => {
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" data-testid="test-1" />,
    );
    expect(getByRole("radio")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <InternalRadio
        id="radio-1"
        name="internal-radio-group"
        value="Option 1"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("has a role of radio", () => {
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" />,
    );

    expect(getByRole("radio")).toBeInTheDocument();
  });

  it("can be checked and unchecked through clicking", () => {
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" />,
    );
    getByRole("radio").click();
    expect(getByRole("radio")).toBeChecked();
  });

  it("can be clicked through its ref", () => {
    const ref = createRef<HTMLInputElement>();
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" ref={ref} />,
    );
    ref.current?.click();
    expect(getByRole("radio")).toBeChecked();
  });

  it("is unselectable when disabled", () => {
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" disabled />,
    );
    getByRole("radio").click();
    expect(getByRole("radio")).not.toBeChecked();
  });

  it("calls onChange method when changed", () => {
    const onChange = jest.fn();
    const { getAllByRole } = renderWithTheme(
      <div>
        <InternalRadio id="radio-1" value="Option 1" onChange={onChange} />
        <InternalRadio id="radio-1" value="Option 2" onChange={onChange} />
      </div>,
    );
    const radios = getAllByRole("radio");
    radios[0]!.click();
    // Note this doesn't trigger as its already selected
    radios[0]!.click();
    radios[1]!.click();
    expect(onChange).toHaveBeenCalledTimes(2);
  });

  it("calls onFocus and onBlur with focus and blur", () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { getByRole } = renderWithTheme(
      <InternalRadio
        id="radio-1"
        value="Option 1"
        onFocus={onFocus}
        onBlur={onBlur}
      />,
    );
    getByRole("radio").focus();
    getByRole("radio").blur();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledTimes(1);
  });

  it("calls onHovered with mouseenter and mouseleave", () => {
    const onHovered = jest.fn();
    const { getByRole } = renderWithTheme(
      <InternalRadio id="radio-1" value="Option 1" onHovered={onHovered} />,
    );
    fireEvent.mouseEnter(getByRole("radio"));
    fireEvent.mouseLeave(getByRole("radio"));
    expect(onHovered).toHaveBeenCalledTimes(1);
  });
});
