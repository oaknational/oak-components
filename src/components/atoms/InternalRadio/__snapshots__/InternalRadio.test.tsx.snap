// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalRadio matches snapshot 1`] = `
.c0 {
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  margin: 0;
  outline: none;
  border: 0.125rem solid;
  border-color: #222222;
  border-radius: 0.125rem;
}

.c0:checked {
  background: #222222;
}

.c0:disabled {
  pointer-events: none;
}

@media (hover:hover) {
  .c0:hover:not(:disabled) {
    border-color: #222222;
  }
}

<input
  className="c0"
  id="radio-1"
  name="internal-radio-group"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="radio"
  value="Option 1"
/>
`;
